### PR TITLE
DTSRD-59: Upgrade spring boot for rd-commondata-api

### PR DESCRIPTION
**Pull request by kiran-yenigala-hmcts**
**Before creating a pull request make sure that:**

- [ ] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [ ] tests have been updated / new tests has been added (if needed)

Please remove this line and everything above and fill the following sections:

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/cdapi/config/LaunchDarklyConfiguration.java [hunk 1]
```diff
@@ -1,21 +1,13 @@
...
 import org.springframework.context.annotation.Configuration;
 import org.springframework.web.servlet.config.annotation.InterceptorRegistry;
 import org.springframework.web.servlet.config.annotation.WebMvcConfigurer;
 
 @Configuration
 public class LaunchDarklyConfiguration implements WebMvcConfigurer {
 
-    @Bean
-    public LDClient ldClient(@Value("${launchdarkly.sdk.key}") String sdkKey) {
-        return new LDClient(sdkKey);
```
Comment (JackBraceyCGI) [inline src/main/java/uk/gov/hmcts/reform/cdapi/config/LaunchDarklyConfiguration.java:16]: Is this replaced with something else, surely we'll lose access to feature flags without this?

**Review by lukasz-wolski — COMMENTED** [review]
please see my comments inlined. minor changes proposed

File: build.gradle [hunk 2]
```diff
@@ -10,7 +10,7 @@ plugins {
   id 'jacoco'
   id "org.flywaydb.flyway" version "7.0.2"
   id 'io.spring.dependency-management' version '1.1.0'
-  id 'org.springframework.boot' version '2.7.12'
+  id 'org.springframework.boot' version '3.1.0'
```
Comment (lukasz-wolski) [inline build.gradle:13]: spring 3.1.0 is not compatible with flyway 9.0 according to https://github.com/spring-projects/spring-boot/issues/35721, spring 3.1.1 with fix is available

File: src/main/java/uk/gov/hmcts/reform/cdapi/config/LaunchDarklyConfiguration.java (hunk 1 above)
Comment (kiran-yenigala-hmcts) [reply src/main/java/uk/gov/hmcts/reform/cdapi/config/LaunchDarklyConfiguration.java:16]: uk.gov.hmcts.reform.cdapi.config.LaunchDarklyClientConfig added in this class

File: src/main/java/uk/gov/hmcts/reform/cdapi/exception/handler/GlobalExceptionHandler.java [hunk 3]
```diff
@@ -132,6 +133,22 @@ private static Throwable getRootException(Throwable exception) {
...
     }
 
+    private ResponseEntity<Object> errorDetailsResponseEntity(Exception ex,
+                                                              HttpStatusCode httpStatus,
+                                                              String errorMsg) {
+
+        log.info(HANDLING_EXCEPTION_TEMPLATE, loggingComponentName, ex.getMessage(), ex);
+        ErrorResponse errorDetails = new ErrorResponse(
+            httpStatus.value(),
+            "",
```
[the remark this answers is not in the document]
Comment (kiran-yenigala-hmcts) [reply src/main/java/uk/gov/hmcts/reform/cdapi/exception/handler/GlobalExceptionHandler.java:143]: changed it now..

File: src/main/java/uk/gov/hmcts/reform/cdapi/config/FeatureConditionEvaluation.java [hunk 4]
```diff
@@ -35,9 +33,9 @@ public class FeatureConditionEvaluation implements HandlerInterceptor {
     private final FeatureToggleService featureToggleService;
 
     @Override
-    public boolean preHandle(HttpServletRequest request,
-                             @NotNull HttpServletResponse response, @NotNull Object handler) throws Exception {
-
+    public boolean preHandle(jakarta.servlet.http.HttpServletRequest request,
+                             jakarta.servlet.http.HttpServletResponse response,
+                             Object handler) throws Exception {
```
[the remark this answers is not in the document]
Comment (kiran-yenigala-hmcts) [reply src/main/java/uk/gov/hmcts/reform/cdapi/config/FeatureConditionEvaluation.java:38]: added it now

File: src/main/java/uk/gov/hmcts/reform/cdapi/config/FeatureConditionEvaluation.java [hunk 5]
```diff
@@ -35,9 +33,9 @@ public class FeatureConditionEvaluation implements HandlerInterceptor {
     private final FeatureToggleService featureToggleService;
 
     @Override
-    public boolean preHandle(HttpServletRequest request,
-                             @NotNull HttpServletResponse response, @NotNull Object handler) throws Exception {
-
+    public boolean preHandle(jakarta.servlet.http.HttpServletRequest request,
+                             jakarta.servlet.http.HttpServletResponse response,
```
[the remark this answers is not in the document]
Comment (kiran-yenigala-hmcts) [reply src/main/java/uk/gov/hmcts/reform/cdapi/config/FeatureConditionEvaluation.java:37]: done

File: src/main/java/uk/gov/hmcts/reform/cdapi/exception/handler/GlobalExceptionHandler.java [hunk 6]
```diff
@@ -60,8 +61,8 @@ public ResponseEntity<Object> handleIllegalArgumentException(IllegalArgumentExce
 
     @ExceptionHandler(HttpStatusCodeException.class)
     public ResponseEntity<Object> handleHttpStatusException(HttpStatusCodeException ex) {
-        HttpStatus httpStatus = ex.getStatusCode();
-        return errorDetailsResponseEntity(ex, httpStatus, httpStatus.getReasonPhrase());
+        HttpStatusCode httpStatus = ex.getStatusCode();
+        return errorDetailsResponseEntity(ex, httpStatus, "");
```
[the remark this answers is not in the document]
Comment (kiran-yenigala-hmcts) [reply src/main/java/uk/gov/hmcts/reform/cdapi/exception/handler/GlobalExceptionHandler.java:65]: changed now

File: src/main/java/uk/gov/hmcts/reform/cdapi/domain/CaseFlag.java [hunk 7]
```diff
@@ -13,6 +14,7 @@
 @NoArgsConstructor
 @AllArgsConstructor
 @Setter
+@JsonIgnoreProperties(ignoreUnknown = true)
```
[the remark this answers is not in the document]
Comment (kiran-yenigala-hmcts) [reply src/main/java/uk/gov/hmcts/reform/cdapi/domain/CaseFlag.java:17]: better to do that if we have new properties defined

File: build.gradle (hunk 2 above)
Comment (kiran-yenigala-hmcts) [reply build.gradle:13]: changed to 3.1.1

File: src/integrationTest/java/uk/gov/hmcts/reform/cdapi/config/SwaggerPublisherTest.java [hunk 8]
```diff
@@ -34,10 +32,7 @@ class SwaggerPublisherTest extends CdAuthorizationEnabledIntegrationTest {
 
     @BeforeEach
     public void setUp() {
-        WebRequestTrackingFilter filter = new WebRequestTrackingFilter();
```
[the remark this answers is not in the document]
Comment (JackBraceyCGI) [reply src/integrationTest/java/uk/gov/hmcts/reform/cdapi/config/SwaggerPublisherTest.java:37]: Did we figure this out, as user profile has the same issue, this just doesn't feel like the best way to resolve it in my mind

[the remark this answers is not in the document]
Comment (kiran-yenigala-hmcts) [reply src/integrationTest/java/uk/gov/hmcts/reform/cdapi/config/SwaggerPublisherTest.java:37]: https://github.com/microsoft/ApplicationInsights-Java-Repros/tree/main/SpringBoot3xSupport

See this for your reference, we can support application insights in spring boot 3 differently than earlier versions

**Review by JackBraceyCGI — APPROVED** [review]

File: src/main/java/uk/gov/hmcts/reform/cdapi/config/SecurityConfiguration.java [hunk 9]
```diff
@@ -61,7 +63,7 @@ public void setAnonymousPaths(List<String> anonymousPaths) {
 
     @Bean
     public WebSecurityCustomizer webSecurityCustomizer() {
-        return web -> web.ignoring().antMatchers(anonymousPaths.toArray(String[]::new));
+        return web -> web.ignoring().requestMatchers(anonymousPaths.toArray(String[]::new));
```
[the remark this answers is not in the document]
Comment (lukasz-wolski) [reply src/main/java/uk/gov/hmcts/reform/cdapi/config/SecurityConfiguration.java:65]: AntPathRequestMatcher instances are created in AbstractRequestMatcherRegistry, closing

**Review by lukasz-wolski — APPROVED** [review]

File: config/owasp/suppressions.xml [hunk 10]
```diff
@@ -14,5 +14,14 @@ file name: launchdarkly-java-server-sdk-5.10.2.jar (shaded: org.yaml:snakeyaml:1
   <suppress>
     <notes>Suppressing false positive, CVE relevant for json-smart up to 2.4.9, but newer version is used</notes>
     <cve>CVE-2023-1370</cve>
+    <cve>CVE-2023-34053</cve>
+    <cve>CVE-2023-34042</cve>
+    <cve>CVE-2023-34055</cve>
+    <cve>CVE-2023-44487</cve>
+    <cve>CVE-2023-33202</cve>
+    <cve>CVE-2023-42795</cve>
```
Comment (lukasz-wolski) [inline config/owasp/suppressions.xml:22]: Is suppression still necessary? CVE has been addressed by https://github.com/hmcts/rd-commondata-api/pull/369

File: config/owasp/suppressions.xml [hunk 11]
```diff
@@ -14,5 +14,14 @@ file name: launchdarkly-java-server-sdk-5.10.2.jar (shaded: org.yaml:snakeyaml:1
...
+    <cve>CVE-2023-46589</cve>
```
Comment (lukasz-wolski) [inline config/owasp/suppressions.xml:23]: Is suppression still necessary? CVE has been addressed by https://github.com/hmcts/rd-commondata-api/pull/412

File: config/owasp/suppressions.xml [hunk 12]
```diff
@@ -14,5 +14,14 @@ file name: launchdarkly-java-server-sdk-5.10.2.jar (shaded: org.yaml:snakeyaml:1
...
+    <cve>CVE-2023-6378</cve>
```
Comment (lukasz-wolski) [inline config/owasp/suppressions.xml:24]: Is suppression still necessary? CVE has been addressed by https://github.com/hmcts/rd-commondata-api/pull/413

File: config/owasp/suppressions.xml [hunk 13]
```diff
@@ -14,5 +14,14 @@ file name: launchdarkly-java-server-sdk-5.10.2.jar (shaded: org.yaml:snakeyaml:1
...
+    <cve>CVE-2023-45648</cve>
```
Comment (lukasz-wolski) [inline config/owasp/suppressions.xml:25]: Is suppression still necessary? CVE has been addressed by https://github.com/hmcts/rd-commondata-api/pull/369

File: src/main/resources/testcontainers.properties [hunk 14]
```diff
@@ -1 +1,2 @@
+image.substitutor=uk.gov.hmcts.reform.cdapi.util.ContainerImageNameSubstitutor
```
Comment (lukasz-wolski) [inline src/main/resources/testcontainers.properties:1]: property below configures an image name prefix for testcontainers. Is this property and implementation class still needed?

**Review by lukasz-wolski — APPROVED** [review]

File: src/main/resources/testcontainers.properties (hunk 14 above)
Comment (kiran-yenigala-hmcts) [reply src/main/resources/testcontainers.properties:1]: removed this

File: config/owasp/suppressions.xml (hunk 13 above)
Comment (kiran-yenigala-hmcts) [reply config/owasp/suppressions.xml:25]: removed all suppressions

**Review by lukasz-wolski — APPROVED** [review]

**Review by rajkatla-hmcts — APPROVED** [review]

**Pull request merged**